### PR TITLE
Storage Server Version Triage

### PR DESCRIPTION
**Pull request by neuroscr**
- builds a pool per version (and a reverse map)
- filters pools with 2.0.2 or greater for file server proxy
- mark as bad removes from random and version pools
- increased start up retries from 1s to 10s (as this PR adds 10-20s delay to proxy services while we get versions and sorts in the usable buckets)
- takes about 10mins to get versions from 980 sndoes

### CODE REVIEW

**Comment by Mikunj** [comment]
> increased start up retries from 1s to 10s (as this PR adds 10-20s delay to proxy services while we get versions and sorts in the usable buckets)

What would this mean in terms of startup time and receiving messages?

> takes about 10mins to get versions from 980 snodes

Wouldn't it be better to query snodes until we have X amount that are of a certain version, and keep doing it in the background?

**Comment by neuroscr** [comment]
> What would this mean in terms of startup time and receiving messages?
Only should really affect proxy requests (RSS, version check, attachments, device mapping fetching)

> Wouldn't it be better to query snodes until we have X amount that are of a certain version, and keep doing it in the background?
The version fetching is backgrounded for the most part.

**Comment by neuroscr** [comment]
Notes for V2:
- sql storage
- store version in random pool and just filter it

**Comment by neuroscr** [comment]
Refined patch to only affect file upload (attachments/avatars) proxy requests allowing RSS, version check, and device mapping fetching to proceed normally since their content is usually under 1mb.

Also handled some edge cases in the markRandomSNodeBad when we haven't got it's version yet, so we retry, so we can mark it bad if/when we do get it.

File: js/modules/loki_snode_api.js [hunk 1]
```diff
@@ -402,15 +525,44 @@ class LokiSnodeAPI {
...
+          // we don't know our version yet
+          // and if we're offline, we'll likely not get it until it restarts if it does...
+          log.warn(
+            'loki_snode:::markRandomNodeUnreachable - No version for snode',
+            `${snode.ip}:${snode.port}`,
+            'retrying in 10s'
+          );
+        }
+        // make sure we don't retry past 15 mins (10s * 100 ~ 1000s)
+        if (options.retries < 100) {
```
Comment (Mikunj) [inline js/modules/loki_snode_api.js]:
```suggestion
        const retries = option.retries || 0;
        if (retries < 100) {
```
we should do this in the case that we forgot to specify retry.

File: js/modules/loki_snode_api.js [hunk 2]
```diff
@@ -249,11 +264,85 @@ class LokiSnodeAPI {
...
 
+  // use nodes that support more than 1mb
+  async getRandomProxySnodeAddress() {
+    /* resolve random snode */
+    if (this.randomSnodePool.length === 0) {
+      // allow exceptions to pass through upwards
+      await this.refreshRandomPool();
+    }
+    if (this.randomSnodePool.length === 0) {
+      throw new window.textsecure.SeedNodeError('Invalid seed node response');
```
Comment (msgmaxim) [inline js/modules/loki_snode_api.js]: Should we throw this in `refreshRandomPool()` instead?

Comment (neuroscr) [reply js/modules/loki_snode_api.js]: Not sure, @Mikunj do you know if this is an issue that we throw inside a promise instead of the outer function?

Comment (msgmaxim) [reply js/modules/loki_snode_api.js]: Throwing inside `async` should reject the promise and `await`ing on rejected promise should just re-throw the original exception.

Comment (neuroscr) [reply js/modules/loki_snode_api.js]: which we catch on line 483 and rethrow

**Comment by neuroscr** [comment]
I really hate how this floods the log with `no version` retries but we can fix up that later

**Pull request merged**